Fix failing grand_canonical sample test and add documentation to samples

jonaslandsgesell opened the pull request:
This PR adds documentation in to the sample files:

* widom_insertion.py
* wang_landau_reaction_ensemble.py
* grand_canonical.py
* reaction_ensemble.py

The PR also fixes partly #3104: The problem with the failing test was that the excess chemical potential did not match the submitted concentration.
I now provide a matching pair of concentration and excess chemical potential.

jonaslandsgesell committed:
add documentation

jonaslandsgesell committed:
fix test

jonaslandsgesell committed:
up

jngrad commented on testsuite/scripts/samples/test_grand_canonical.py:
```diff
@@ -20,7 +20,7 @@
 sample, skipIfMissingFeatures = importlib_wrapper.configure_and_import(
     "@SAMPLES_DIR@/grand_canonical.py", n_int_cycles=51, n_int_steps=5,
-    warm_n_times=10, warm_steps=50, cmd_arguments=[0.01, 0.01])
+    warm_n_times=10, warm_steps=50, cmd_arguments=[8.523659461370000200e-04, -3.365753246398797693e-01])
```
how many significant figures do we really need here?

jonaslandsgesell replied:
8.523659461370000200e-04 was the concentration I used for determining the excess chemical potential. This is therefore exact. 
The value for the excess chemical potential has a standard error of  +/- 1.590969027075880404e-03 therefore I cropped the result to the third digit after the point: -0.336

jngrad replied:
you're providing a concentration with 19 significant figures... either WLRE relies on chaos theory, or you're not familiar with Avogadro's constant :) please trim that down, espresso runs on multiple CPU architectures with different machine precision, your value doesn't even fit inside a `long double` on a 64bit machine

jngrad commented on samples/wang_landau_reaction_ensemble.py:
```diff
@@ -17,7 +17,9 @@
-This sample simulates the Wang-Landau Reaction Ensemble for a harmonic bond.
+This example script simulates two reacting monomers which are bonded via a harmonic potential.
+The script aborts as soon as the abortion criterion in the Wang-Landau algorithm is met.
+The Wang-Landau potential which is updated during the simulation is written to the file WL_potential_out.dat.
```
Just to check my understanding: we have a neutral monomer (type 3) and a base monomer (type A-) connected by a bond, and two protons (type H+) to set the pH. The WLRE algorithm changes the base monomer into its conjugated acid (type AH) and back using MC moves and stores the degree of association ñ and the chemical potential µ to two files.

jonaslandsgesell replied:
The situtation is the following: We have two bonded monomers. These can change their type. They are either of type A- or of type HA. When they are of type A- they have released the corresponding H+ ion. The pH would have to be determined through the reweighted H+ concentration after the simulation is over.
The Wang-Landau algorithm is a histogram reweighting algorithm and drives the system out of equilibrium via applying a Wang-Landau potential. This is not directly connected to the chemical potential. The Wang-Landau potential either represents the canonical partition function, the density of states or the a mixture of both, depending on the acceptance probabilities you impose on the system.
A detailed explaination can be found in https://pubs.acs.org/doi/full/10.1021/acs.jctc.6b00791 .

jngrad replied:
Thanks!

jonaslandsgesell committed:
add doc

jonaslandsgesell committed:
wording

jonaslandsgesell committed:
fmt

jonaslandsgesell committed:
truncate and add comment to how one obtains the magic parameters

jonaslandsgesell committed:
fmt

jngrad committed:
Cleanup minimization code

jngrad committed:
Increase test tolerance

Running this test 5000 times resulted in 65 occurences of a
deviation larger than 10%, with a maximal deviation of 13.6%, a
mean of 3.3% and a median of 2.9%. Setting the test threshold to
15% should significantly reduce the frequency of random failures.

jngrad committed:
Document origin of hardcoded values

jngrad committed:
Use argparse module

jngrad committed:
Remove repetitions in the documentation

jngrad committed:
Re-shuffle Doxygen blocks

jngrad committed:
Update documentation

jngrad committed:
Formatting and numpydoc style

jngrad committed:
Fix typos and Sphinx/Doxygen issues

jonaslandsgesell reviewed (commented):
Looks good to me

jngrad commented:
bors r+
